File: src/models/schedule.py
"""Schedule models for Feature 011 — Proactive Assistant."""

from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field


class TaskType(str, Enum):
    """Scheduled task type."""

    ONE_TIME = "one_time"
    RECURRING = "recurring"


class TaskStatus(str, Enum):
    """Scheduled task status."""

    ACTIVE = "active"
    PAUSED = "paused"
    CANCELLED = "cancelled"
    COMPLETED = "completed"


class TaskSource(str, Enum):
    """Who created the task."""

    USER = "user"
    AGENT = "agent"


class RunStatus(str, Enum):
    """Task run execution status."""

    RUNNING = "running"
    SUCCESS = "success"
    FAILED = "failed"
    RETRYING = "retrying"


class ScheduledTask(BaseModel):
    """A one-time or recurring scheduled task."""

    id: UUID
    user_id: UUID
    name: str
    description: Optional[str] = None
    task_type: TaskType
    schedule_cron: Optional[str] = None
    scheduled_at: Optional[datetime] = None
    timezone: str = "UTC"
    tool_name: str
    tool_args: dict = Field(default_factory=dict)
    prompt_template: str
    status: TaskStatus = TaskStatus.ACTIVE
    source: TaskSource = TaskSource.USER
    next_run_at: Optional[datetime] = None
    last_run_at: Optional[datetime] = None
    run_count: int = 0
    fail_count: int = 0
    max_retries: int = 3
    created_at: datetime
    updated_at: datetime


class TaskRun(BaseModel):
    """A single execution of a scheduled task."""

    id: UUID
    task_id: UUID
    started_at: datetime
    completed_at: Optional[datetime] = None
    status: RunStatus = RunStatus.RUNNING
    result: Optional[str] = None
    error: Optional[str] = None
    notification_id: Optional[UUID] = None
    retry_count: int = 0
    duration_ms: Optional[int] = None

File: frontend/src/types/schedule.ts
export type TaskType = "one_time" | "recurring";

export type TaskStatus = "active" | "paused" | "cancelled" | "completed";

export type TaskSource = "user" | "agent";

export interface ScheduledTask {
  id: string;
  name: string;
  description: string | null;
  task_type: TaskType;
  schedule_cron: string | null;
  scheduled_at: string | null;
  timezone: string;
  tool_name: string;
  status: TaskStatus;
  source: TaskSource;
  next_run_at: string | null;
  last_run_at: string | null;
  run_count: number;
  fail_count: number;
  created_at: string;
}

export interface TaskRun {
  id: string;
  task_id: string;
  started_at: string;
  completed_at: string | null;
  status: "running" | "success" | "failed" | "retrying";
  result: string | null;
  error: string | null;
  notification_id: string | null;
  retry_count: number;
  duration_ms: number | null;
}

export interface PaginatedSchedules {
  items: ScheduledTask[];
  total: number;
  limit: number;
  offset: number;
}
